plugins/outputs/groundwork/groundwork.go

```diff
@@ -0,0 +1,363 @@
+
+	return nil
+}
+
+func (g *Groundwork) Close() error {
+	formValues := map[string]string{
+		"gwos-app-name":  "telegraf",
+		"gwos-api-token": g.authToken,
+	}
+
+	headers := map[string]string{
+		"Content-Type": "application/x-www-form-urlencoded",
+		"User-Agent":   internal.ProductToken(),
+	}
+
+	if _, _, err := clients.SendRequest(http.MethodPost, g.Server+logoutURL, headers, formValues, nil); err != nil {
+		return fmt.Errorf("could not log out at %s: %v", g.Server+logoutURL, err)
+	}
+
+	return nil
+}
+
+func (g *Groundwork) Write(metrics []telegraf.Metric) error {
+	resourceToServicesMap := make(map[string][]transit.DynamicMonitoredService)
+	for _, metric := range metrics {
+		resource, service := g.parseMetric(metric)
+		resourceToServicesMap[resource] = append(resourceToServicesMap[resource], service)
+	}
+
+	var resources []transit.DynamicMonitoredResource
+	for resourceName, services := range resourceToServicesMap {
+		resources = append(resources, transit.DynamicMonitoredResource{
+			BaseResource: transit.BaseResource{
+				BaseTransitData: transit.BaseTransitData{
+					Name: resourceName,
+					Type: transit.Host,
+				},
+			},
+			Status:        transit.HostUp,
+			LastCheckTime: transit.NewTimestamp(),
+			Services:      services,
+		})
+	}
+
+	traceToken, _ := uuid.GenerateUUID()
+	requestJSON, err := json.Marshal(transit.DynamicResourcesWithServicesRequest{
+		Context: &transit.TracerContext{
+			AppType:    "TELEGRAF",
+			AgentID:    g.AgentID,
+			TraceToken: traceToken,
+			TimeStamp:  transit.NewTimestamp(),
+			Version:    transit.ModelVersion,
+		},
+		Resources: resources,
+		Groups:    nil,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	headers := map[string]string{
+		"GWOS-APP-NAME":  "telegraf",
+		"GWOS-API-TOKEN": g.authToken,
+		"Content-Type":   "application/json",
+		"Accept":         "application/json",
+		"User-Agent":     internal.ProductToken(),
+	}
+
+	statusCode, body, err := clients.SendRequest(http.MethodPost, g.Server+defaultMonitoringRoute, headers, nil, requestJSON)
+	if err != nil {
+		return fmt.Errorf("error while sending: %v", err)
+	}
+
+	/* Re-login mechanism */
+	if statusCode == 401 {
+		if err = g.Connect(); err != nil {
+			return fmt.Errorf("re-login failed: %v", err)
+		}
+		headers["GWOS-API-TOKEN"] = g.authToken
+		statusCode, body, err = clients.SendRequest(http.MethodPost, g.Server+defaultMonitoringRoute, headers, nil, requestJSON)
+		if err != nil {
+			return fmt.Errorf("error while sending: %v", err)
+		}
+	}
+
+	if statusCode != 200 {
+		return fmt.Errorf("HTTP request failed. [Status code]: %d, [Response]: %s", statusCode, string(body))
+	}
+
+	return nil
+}
+
+func (g *Groundwork) Description() string {
+	return "Send telegraf metrics to GroundWork Monitor"
+}
+
+func init() {
+	outputs.Add("groundwork", func() telegraf.Output {
+		return &Groundwork{
+			ResourceTag:         "host",
+			DefaultHost:         "telegraf",
+			DefaultServiceState: string(transit.ServiceOk),
```

Comment:
Please set those in `Init()` as it makes code easier...

Reply:
https://github.com/influxdata/telegraf/pull/9891#discussion_r755796711

Reply:
And my reply. ;-) Let's discuss there and I will mark this as resolved.

Reply:
The plugin can't work without url, agent_id, username and password, but it can easily work with default values for other configuration. I find it okay to check the string and set the default if it's empty 👌